sphynx/lynxkite-sphynx/networkit_compute_double_edge_attribute.go

```diff
@@ -0,0 +1,59 @@
+// All NetworKit ops that compute a Double edge attribute on a graph.
+package main
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"runtime/debug"
+
+	"github.com/lynxkite/lynxkite/sphynx/networkit"
+)
+
+func init() {
+	operationRepository["NetworKitComputeDoubleEdgeAttribute"] = Operation{
+		execute: func(ea *EntityAccessor) (err error) {
+			defer func() {
+				if e := recover(); e != nil {
+					err = fmt.Errorf("%v", e)
+					log.Printf("%v\n%v", e, string(debug.Stack()))
+				}
+			}()
+			vs := ea.getVertexSet("vs")
+			es := ea.getEdgeBundle("es")
+			weight := ea.getDoubleAttributeOpt("weight")
+			o := &NetworKitOptions{ea.GetMapParam("options")}
+			seed := uint64(1)
+			if s, exists := o.Options["seed"]; exists {
+				seed = uint64(s.(float64))
+			}
+			networkit.SetSeed(seed, true)
+			networkit.SetThreadsFromEnv()
+			// The caller can set "directed" to false to create an undirected graph.
+			g := ToNetworKit(vs, es, weight, o.Options["directed"] != false)
+			defer networkit.DeleteGraph(g)
+			g.IndexEdges()
+			var result networkit.DoubleVector
+			switch ea.GetStringParam("op") {
+			case "ForestFireScore":
+				c := networkit.NewForestFireScore(
+					g, o.Double("spread_prob"), o.Double("burn_ratio"))
+				defer networkit.DeleteForestFireScore(c)
+				c.Run()
+				result = c.Scores()
+			}
+			attr := &DoubleAttribute{
+				Values:  make([]float64, len(es.Src)),
+				Defined: make([]bool, len(es.Src)),
+			}
+			for i := range attr.Values {
+				// The NetworKit edge IDs don't correspond to the Sphynx edge IDs.
+				// TODO: Can we do better than mapping the results back by src/dst?
```

Comment:
This sucks.

NetworKit edges doesn't normally have IDs because they don't really need to be identified. Except in these edge scoring algorithms. So they require you to run `indexEdges()` first. `indexEdges()` just hands out edge IDs sequentially from 0. Ideally we would have the same edge IDs in NetworKit as in Sphynx, so the returned scores could be applied directly. But some other mapping could work too. I didn't really find anything that would work, and now have to loop up the edges by src/dst.

One idea is to put the Sphynx ID into the weight when I build the graph. Then call `indexEdges()` and we can read out the mapping. But I can't find a way to then replace these fake weights with the real weights. And anyway it would be pretty ugly.

Let me know if you have better ideas.

Reply:
Why is this current solution so bad? Are you worried that `g.EdgeId` is slow?

Somewhat related: what's the deal with parallel edges? As far as it seems to me now, NetworKit does not care for parallel edges. So there can be only one weight for an (a, b) pair and there will only be one result as well. Do we handle this correctly when we convert the weights for NK? Because of this, we cannot hope for a 1 to 1 correspondence between our ids and networKit ids, can we?

Reply:
> Why is this current solution so bad? Are you worried that `g.EdgeId` is slow?

Yes. It looks up `u` in `outEdges` (O(1)) and then scans through the outgoing edge list to find `v` (O(`outEdges[u].size`)). Knowing its index it can look up the ID in `outEdgeIds` (O(1)).

```c++
    std::vector<std::vector<node>> outEdges;
    std::vector<std::vector<edgeid>> outEdgeIds;
    std::vector<std::vector<edgeweight>> outEdgeWeights;
```

> Somewhat related: what's the deal with parallel edges?

Yeah, this will mix up the parallel edges. We will output the score of the "first" edge for all parallel edges.

NetworKit can represent parallel edges. You just have the same neighbor twice in `outEdges[u]`. Some algorithms may not handle them well, as they may look up edge weights by src/dst which will just return the first weight.

Reply:
I don't seem to find it, but in case I'm just missing it: is there a (hopefully O(1)) way to get src and dst for a given NK edge id? If yes, then worse case you can just use this to build a (src, dst) -> NK Id map on your end and use this to avoid the O(outdegree) cost. But even if possible, I'd defer that to a separate issue for now. I don't have any better idea than this unfortunately.

Reply:
> I don't seem to find it, but in case I'm just missing it: is there a (hopefully O(1)) way to get src and dst for a given NK edge id? If yes, then worse case you can just use this to build a (src, dst) -> NK Id map on your end and use this to avoid the O(outdegree) cost. But even if possible, I'd defer that to a separate issue for now. I don't have any better idea than this unfortunately.

Not exactly from ID to src/dst, but we can look up both together! This should work! Thanks. I think I'll just do it now. I won't remember anything by next year.

```c++
    std::pair<node, edgeid> getIthNeighborWithId(node u, index i) const {
```

Reply:
This API looks like a real pleasure to work with. :) But OK, if you do it, that's great!